File: Problem1_SortColors.cpp

// Time Complexity : O(n) where n is number of elements
// Space Complexity : O(1) we are not using extra space
// Did this code successfully run on Leetcode :yes
// Any problem you faced while coding this : no


// Your code here along with comments explaining your approach


#include<vector>
#include<bits/stdc++.h>
#include<stdlib.h>
#include<cmath>
#include<time.h>
#include<iostream>

using namespace std;

class Solution {
public:
    void sortColors(vector<int>& nums) {
        
        if(nums.empty()) return;
        int low=0;
        int mid=0;
        int high=nums.size()-1;
        
        while(mid<=high){
            if(nums[mid]==0){
                swap(nums[low],nums[mid]);
                low++;
                mid++;
            }else if(nums[mid]==1){
                mid++;
            }else{
                swap(nums[high],nums[mid]);
                high--;
            }
        }
        
        
        
        
    }
};

int main(){
    Solution a;
    vector<int>  nums={0,0,1,1,1,1,2};
    //int target=5;
    a.sortColors(nums);
    for (int x : nums) 
         cout << x << " "; 

    cout<<endl;

    // can either do this below for answer in true false 
    // cout.setf(std::ios::boolalpha);
    // cout <<  b << endl;

    // or

    // cout <<  boolalpha << b << endl;

}

File: Problem1_SortColors.java
// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

// Using three pointer approach, take low and mid at 0th index and right pointer at N-1 index
// if mid==2, swap it to the right pointer, if mid==0, swap it to the low pointer
// after multiple iterations, when m > r the array is sorted 

public class Problem1_SortColors {
    public void sortColors(int[] nums) {
        int n=nums.length;
        int l=0, m=0, r=n-1;
        while(m<=r){
            if(nums[m]==2){
                swap(nums, m, r);
                r--;
            }else if(nums[m]==0){
                swap(nums, l, m);
                l++;
                m++;
            }else{
                m++;
            }
        }
    }
    private void swap(int[] nums, int i, int j){
        int temp=nums[i];
        nums[i]=nums[j];
        nums[j]=temp;
    }
}
